Get platform independent text alignment value

klaaspieter opened the pull request:
NSTextAlignment uses different numerical values for different
architectures. Using the raw value means we incorrectly report text
alignment on M1 machines.

This fixes the issue by using `MSTextAlignmentConverter` used internally
by Sketch.

connects https://github.com/sketch-hq/Sketch/issues/39019

Closes https://github.com/sketch-hq/SketchAPI/issues/860

klaaspieter committed:
Get platform independent text alignment value

NSTextAlignment uses different numerical values for different
architectures. Using the raw value means we incorrectly report text
alignment on M1 machines.

This fixes the issue by using `MSTextAlignmentConverter` used internally
by Sketch.

Fixes https://github.com/sketch-hq/Sketch/issues/39019

Depends on this Sketch PR https://github.com/sketch-hq/Sketch/pull/42930

lrossi commented on Source/dom/style/Text.js:
```diff
@@ -175,7 +175,10 @@ export function defineTextStyleProperties(Style) {
       }
 
-      const raw = paragraphStyle.alignment()
+      // eslint-disable-next-line
```
Another option is to add `MSTextAlignmentConverter` to the `.eslintrc`, which seems to be how Sketch's types are handled in the project.

lrossi commented on Source/dom/style/Text.js:
```diff
@@ -175,7 +175,10 @@ export function defineTextStyleProperties(Style) {
 
-      const raw = paragraphStyle.alignment()
+      // eslint-disable-next-line
+      let raw = MSTextAlignmentConverter.archiveNSTextAlignment(
```
May be worth adding a comment explaining why this is necessary, given the method name doesn't particularly help 🙂

lrossi commented:
@klaaspieter great job, only a couple of minor comments 👍

lrossi commented:
@klaaspieter all good and tested on M1, though it occurred to me that we need to do the reverse in the setter too 🙃

klaaspieter committed:
Use platform independent text alignment in setter

klaaspieter commented:
@lrossi added it to the setter as well. Can you take another look and make sure it works on M1?

lrossi commented on Source/dom/style/Text.js:
```diff
@@ -184,7 +188,12 @@ export function defineTextStyleProperties(Style) {
+        // NSTextAlignment is represented differently on Intel and M1 chips
+        const translated =
+          TextAlignmentMap[
+            MSTextAlignmentConverter.archiveNSTextAlignment(mode)
```
@klaaspieter it doesn't seem to be working on M1: `mode` appears to be a string here (e.g. `"center"`), maybe we need to pass `TextAlignmentMap[mode]` to `archiveNSTextAlignment` instead? (The other way around)

klaaspieter replied:
Should be fixed now 🤞🏻

klaaspieter committed:
Use rawValue to get platform independent value

The API accepts both `NSTextAlignmentCenter` (which is the integer 2) or
`'center'` as an argument to `style.alignment`. We need to find the raw
numeric value before we can get the platform independent value.

The code assumes that whenever we cannot find the raw value whatever the
user passed in will be a correct value. This is what the previous
implementation did as well.

klaaspieter committed:
Only allow valid alignments

Previously the Mac app would ignore invalid alignment values.
Unfortunately `archiveNSTextAlignment` crashes so we need to make sure
we only give it valid values.

klaaspieter commented:
When testing the setter please test both `center` and `NSTextAlignmentCenter` variants because it took quite some time to get both to work 😓

lrossi reviewed (approved):
👌
